tools/git-utils.sh

```diff
@@ -0,0 +1,80 @@
+# doesn't do anything on its own.  must be sourced.
+
+GITHUB_BASE="${GITHUB_BASE:-cockpit-project/cockpit}"
+GITHUB_REPOSITORY="${GITHUB_BASE%/*}/${GITHUB_REPO}"
+HTTPS_REMOTE="https://github.com/${GITHUB_REPOSITORY}"
+SSH_REMOTE="https://github.com/${GITHUB_REPOSITORY}"
+
+CACHE_DIR="${XDG_CACHE_HOME-${HOME}/.cache}/cockpit-dev/${GITHUB_REPOSITORY}.git"
+
+if [ "${V-}" = 0 ]; then
+    message() { printf "  %-8s %s\n" "$1" "$2"; }
+    quiet='--quiet'
+else
+    message() { :; }
+    quiet=''
+fi
+
+get_index_gitlink() {
+    if ! git ls-files -s "$1" | egrep -o '\<[[:xdigit:]]{40}\>'; then
+        echo "*** couldn't read gitlink for file $1 from the index" >&2
+        exit 1
+    fi
+}
+
+init_cache() {
+    if [ ! -d "${CACHE_DIR}" ]; then
+        message INIT "${CACHE_DIR}"
+        mkdir -p "${CACHE_DIR}"
+        git init --bare --template='' ${quiet} "${CACHE_DIR}"
+        git --git-dir "${CACHE_DIR}" remote add origin "${HTTPS_REMOTE}"
+    fi
+}
+
+check_ref() {
+    git --git-dir "${CACHE_DIR}" rev-parse --verify --quiet "$1" >/dev/null
+}
+
+fetch_sha_to_cache() {
+    sha="$1"
+
+    # No "offline mode" here: we either have the commit, or we don't
+    if ! check_ref "${sha}"; then
+        init_cache
+        message FETCH "${CLONE_TARGET}  [ref: ${sha}]"
+        git --git-dir "${CACHE_DIR}" fetch --no-tags ${quiet} origin "${sha}"
+        # tag it to keep it from being GC'd.
+        git --git-dir "${CACHE_DIR}" tag "sha-${sha}" "${sha}"
+    fi
+}
+
+fetch_to_cache() {
```

Comment:
I added some comments to better explain the difference, but I prefer to keep the split as it is.  It makes sense to me to treat these two cases separately.  Originally, I tried to implement that as one common function, but the two cases are really different.

I want to prevent GC because in the case where we use this function (fetching commit IDs) we currently don't have plans to ever remove items from the upstream cache repository.  If we do want to implement pruning, we should do that separately.

Reply:
Sure, I wasn't talking about the repo on github, but the local ~/.cache repo. Fetching would/should never actually push any tag to origin, so that tagging here should at most influence the local repo?